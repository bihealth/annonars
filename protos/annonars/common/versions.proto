syntax = "proto3";

package annonars.common.versions;

// Source that data was created from.
message CreatedFrom {
    // The name of the data source.
    string name = 1;
    // The version of the data source.
    string version = 2;
}

// Version specification.
message VersionSpec {
    // Identifier of the data.
    string identifier = 1;
    // Title of the data.
    string title = 2;
    // Creator of the data.
    string creator = 3;
    // Contributors of the data.
    repeated string contributor = 4;
    // Format of the data.
    string format = 5;
    // Date of the data.
    string date = 6;
    // Version of the data.
    string version = 7;
    // Optional genome release.
    optional string genome_release = 8;
    // Data description.
    string description = 9;
    // Data source.
    repeated string source = 10;
    // Created from information.
    repeated CreatedFrom created_from = 11;
}
