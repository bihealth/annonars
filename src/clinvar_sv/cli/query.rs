//! Query of ClinVar SV annonars data.

use std::{io::Write, sync::Arc};

use bio::data_structures::interval_tree::ArrayBackedIntervalTree;
use prost::Message;

use crate::common::{self, cli::extract_chrom, spdi};

/// Argument group for specifying accession or range.
#[derive(clap::Args, Debug, Clone, Default)]
#[group(required = true, multiple = false)]
pub struct ArgsQuery {
    /// Specify range to query for.
    #[arg(long, group = "query")]
    pub accession: Option<String>,
    /// Query for all variants.
    #[arg(long, group = "query")]
    pub all: bool,
    /// Specify range to query for.
    #[arg(long, group = "query")]
    pub range: Option<spdi::Range>,
}

/// Command line arguments for `tsv query` sub command.
#[derive(clap::Parser, Debug, Clone)]
#[command(about = "query gnomAD-mtDNA data stored in RocksDB", long_about = None)]
pub struct Args {
    /// Path to RocksDB directory with data.
    #[arg(long)]
    pub path_rocksdb: String,
    /// Name of the column family to import into.
    #[arg(long, default_value = "clinvar_sv")]
    pub cf_name: String,
    /// Mapping from ClinVar RCV to ClinVar VCV.
    #[arg(long, default_value = "clinvar_sv_by_rcv")]
    pub cf_name_by_rcv: String,
    /// Output file (default is stdout == "-").
    #[arg(long, default_value = "-")]
    pub out_file: String,
    /// Output format.
    #[arg(long, default_value = "jsonl")]
    pub out_format: common::cli::OutputFormat,

    /// Variant or position to query for.
    #[command(flatten)]
    pub query: ArgsQuery,
}

/// Meta information as read from database.
#[derive(Debug)]
pub struct Meta {
    /// Genome release of data in database.
    pub genome_release: String,
}

/// Open RocksDb given path and column family name for data and metadata.
pub fn open_rocksdb<P: AsRef<std::path::Path>>(
    path_rocksdb: P,
    cf_data: &str,
    cf_meta: &str,
    cf_by_rcv: &str,
) -> Result<(Arc<rocksdb::DBWithThreadMode<rocksdb::MultiThreaded>>, Meta), anyhow::Error> {
    tracing::info!("Opening RocksDB database ...");
    let before_open = std::time::Instant::now();
    let cf_names = &[cf_meta, cf_data, cf_by_rcv];
    let db = Arc::new(rocksdb::DB::open_cf_for_read_only(
        &rocksdb::Options::default(),
        common::readlink_f(&path_rocksdb)?,
        cf_names,
        true,
    )?);
    tracing::info!("  reading meta information");
    let meta = {
        let cf_meta = db.cf_handle(cf_meta).unwrap();
        let meta_genome_release = String::from_utf8(
            db.get_cf(&cf_meta, "genome-release")?
                .ok_or_else(|| anyhow::anyhow!("missing value meta:genome-release"))?,
        )?;
        Meta {
            genome_release: meta_genome_release,
        }
    };

    tracing::info!("  meta:genome-release = {}", &meta.genome_release);
    tracing::info!(
        "... opening RocksDB database took {:?}",
        before_open.elapsed()
    );

    Ok((db, meta))
}

/// Open RocksDB database from command line arguments.
pub fn open_rocksdb_from_args(
    args: &Args,
) -> Result<(Arc<rocksdb::DBWithThreadMode<rocksdb::MultiThreaded>>, Meta), anyhow::Error> {
    open_rocksdb(
        &args.path_rocksdb,
        &args.cf_name,
        "meta",
        &args.cf_name_by_rcv,
    )
}

fn print_record(
    out_writer: &mut Box<dyn std::io::Write>,
    output_format: common::cli::OutputFormat,
    value: &crate::pbs::annonars::clinvar::v1::sv::Record,
) -> Result<(), anyhow::Error> {
    match output_format {
        common::cli::OutputFormat::Jsonl => {
            writeln!(out_writer, "{}", serde_json::to_string(value)?)?;
        }
    }

    Ok(())
}

/// Query by accession.
pub fn query_for_accession(
    accession: &str,
    db: &rocksdb::DBWithThreadMode<rocksdb::MultiThreaded>,
    cf_data: &Arc<rocksdb::BoundColumnFamily>,
    cf_by_rcv: &Arc<rocksdb::BoundColumnFamily>,
) -> Result<Option<crate::pbs::annonars::clinvar::v1::sv::Record>, anyhow::Error> {
    // Execute query.
    tracing::debug!("accession = {:?}", &accession);
    let vcv = if accession.starts_with("VCV") {
        accession.as_bytes().into()
    } else if accession.starts_with("RCV") {
        db.get_cf(cf_by_rcv, accession.as_bytes())?
            .ok_or_else(|| anyhow::anyhow!("no VCV found for RCV {}", accession))?
    } else {
        anyhow::bail!("Not a valid VCV/RCV accession: {:?}", &accession);
    };
    tracing::debug!("vcv = {:?}", &vcv);

    let raw_value = db
        .get_cf(cf_data, vcv.clone())
        .map_err(|e| anyhow::anyhow!("error while querying for vcv {:?}: {}", vcv, e))?;
    raw_value
        .map(|raw_value| {
            // Decode via prost.
            crate::pbs::annonars::clinvar::v1::sv::Record::decode(&mut std::io::Cursor::new(
                &raw_value,
            ))
            .map_err(|e| anyhow::anyhow!("failed to decode record: {}", e))
        })
        .transpose()
}

/// Query all variants and print to `out_writer`.
fn print_all(
    out_writer: &mut Box<dyn std::io::Write>,
    out_format: common::cli::OutputFormat,
    db: &rocksdb::DBWithThreadMode<rocksdb::MultiThreaded>,
    cf_data: &Arc<rocksdb::BoundColumnFamily>,
) -> Result<(), anyhow::Error> {
    tracing::info!("dumping all records...");

    let mut iter = db.raw_iterator_cf(cf_data);
    iter.seek(b"");
    while iter.valid() {
        if let Some(raw_value) = iter.value() {
            let record = crate::pbs::annonars::clinvar::v1::sv::Record::decode(
                &mut std::io::Cursor::new(&raw_value),
            )
            .map_err(|e| anyhow::anyhow!("failed to decode record: {}", e))?;
            print_record(out_writer, out_format, &record)?;
            iter.next();
        } else {
            break;
        }
    }

    tracing::info!("... done dumping all records");
    Ok(())
}

/// Helper data structure that provides per-chromosome interval trees for querying.
pub struct IntervalTrees {
    /// Per-chromosome interval trees.
    trees: rustc_hash::FxHashMap<String, ArrayBackedIntervalTree<u64, String>>,
    /// Backing RocksDB.
    db: Arc<rocksdb::DBWithThreadMode<rocksdb::MultiThreaded>>,
    /// Name of column family with data.
    cf_data_name: String,
    /// Meta information from database.
    meta: Meta,
}

impl IntervalTrees {
    /// Construct new per-contig interval trees.
    ///
    /// This will read all records from the database and build the interval trees.
    ///
    /// # Arguments
    ///
    /// * `db` - Database to read from.
    /// * `cf_data_name` - Name of column family with data.
    /// * `meta` - Meta information from database.
    ///
    /// # Returns
    ///
    /// * `Self` - New instance.
    ///
    /// # Errors
    ///
    /// * If reading from the database fails.
    pub fn with_db(
        db: Arc<rocksdb::DBWithThreadMode<rocksdb::MultiThreaded>>,
        cf_data_name: &str,
        meta: Meta,
    ) -> Result<Self, anyhow::Error> {
        let cf_data = db.cf_handle(cf_data_name).ok_or_else(|| {
            anyhow::anyhow!("no column family with name {:?} found", cf_data_name)
        })?;
        Ok(Self {
            trees: Self::build_trees(db.clone(), cf_data.clone())?,
            db: db.clone(),
            cf_data_name: cf_data_name.to_string(),
            meta: meta,
        })
    }

    /// Build the interval trees.
    fn build_trees(
        db: Arc<rocksdb::DBWithThreadMode<rocksdb::MultiThreaded>>,
        cf_data: Arc<rocksdb::BoundColumnFamily>,
    ) -> Result<rustc_hash::FxHashMap<String, ArrayBackedIntervalTree<u64, String>>, anyhow::Error>
    {
        let mut result: rustc_hash::FxHashMap<String, ArrayBackedIntervalTree<u64, String>> =
            rustc_hash::FxHashMap::default();

        // Obtain iterator and seek to start.
        let mut iter = db.raw_iterator_cf(&cf_data);
        iter.seek(b"");
        while iter.valid() {
            if let Some(raw_value) = iter.value() {
                let record = crate::pbs::annonars::clinvar::v1::sv::Record::decode(
                    &mut std::io::Cursor::new(&raw_value),
                )
                .map_err(|e| anyhow::anyhow!("failed to decode record: {}", e))?;
                tracing::trace!("iterator at {:?} => {:?}", &iter.key(), &record);

                let crate::pbs::annonars::clinvar::v1::sv::Record {
                    chromosome,
                    start,
                    stop,
                    vcv,
                    ..
                } = record;

                let interval = (start as u64 - 1)..(stop as u64);
                tracing::trace!("contig = {} / {:?} / {}", &chromosome, &interval, &vcv);
                result
                    .entry(chromosome.clone())
                    .or_default()
                    .insert(interval, vcv);
                assert!(result.contains_key(&chromosome));

                iter.next();
            } else {
                break;
            }
        }

        result.values_mut().for_each(|tree| tree.index());

        Ok(result)
    }

    /// Query for a range.
    pub fn query(
        &self,
        range: &spdi::Range,
    ) -> Result<Vec<crate::pbs::annonars::clinvar::v1::sv::Record>, anyhow::Error> {
        let contig = extract_chrom::from_range(&range, Some(&self.meta.genome_release))?;
        let cf_data = self.db.cf_handle(&self.cf_data_name).ok_or_else(|| {
            anyhow::anyhow!("no column family with name {:?} found", &self.cf_data_name)
        })?;
        let interval = (range.start as u64 - 1)..(range.end as u64);
        let mut result = Vec::new();
        if let Some(tree) = self.trees.get(&contig) {
            for entry in tree.find(&interval) {
                if let Some(raw_value) = self.db.get_cf(&cf_data, entry.data().as_bytes())? {
                    let record = crate::pbs::annonars::clinvar::v1::sv::Record::decode(
                        &mut std::io::Cursor::new(&raw_value),
                    )
                    .map_err(|e| anyhow::anyhow!("failed to decode record: {}", e))?;
                    result.push(record);
                }
            }
        } else {
            tracing::warn!("unknown contig: {:?}", &contig);
        }

        Ok(result)
    }
}

/// Implementation of `tsv query` sub command.
pub fn run(common: &common::cli::Args, args: &Args) -> Result<(), anyhow::Error> {
    tracing::info!("Starting 'clinvar-sv query' command");
    tracing::info!("common = {:#?}", &common);
    tracing::info!("args = {:#?}", &args);

    let (db, meta) = open_rocksdb_from_args(args)?;
    let cf_data = db.cf_handle(&args.cf_name).unwrap();
    let cf_by_rcv = db.cf_handle(&args.cf_name_by_rcv).unwrap();

    // Obtain writer to output.
    let mut out_writer = match args.out_file.as_ref() {
        "-" => Box::new(std::io::stdout()) as Box<dyn std::io::Write>,
        out_file => {
            let path = std::path::Path::new(out_file);
            Box::new(std::fs::File::create(path).unwrap()) as Box<dyn std::io::Write>
        }
    };

    tracing::info!("Running query...");
    let before_query = std::time::Instant::now();
    if let Some(accession) = args.query.accession.as_ref() {
        tracing::info!("for accession {}", &accession);
        if let Some(record) = query_for_accession(accession, &db, &cf_data, &cf_by_rcv)? {
            print_record(&mut out_writer, args.out_format, &record)?;
        } else {
            tracing::info!("no record found for accession {:?}", &accession);
        }
    } else if let Some(range) = args.query.range.as_ref() {
        tracing::info!("for range {:?}", &range);
        tracing::info!("Building interval trees...");
        let trees = IntervalTrees::with_db(db.clone(), &args.cf_name, meta)
            .map_err(|e| anyhow::anyhow!("failed to build interval trees: {}", e))?;
        tracing::info!("... done building interval trees");
        tracing::info!("Running query...");
        let records = trees
            .query(&range)
            .map_err(|e| anyhow::anyhow!("failed to query interval trees: {}", e))?;
        for record in &records {
            print_record(&mut out_writer, args.out_format, &record)?;
        }
        tracing::info!("... done running query");
    } else if args.query.all {
        tracing::info!("for all");
        print_all(&mut out_writer, args.out_format, &db, &cf_data)?;
    } else {
        unreachable!();
    }
    tracing::info!("... done querying in {:?}", before_query.elapsed());

    tracing::info!("All done. Have a nice day!");
    Ok(())
}

#[cfg(test)]
mod test {
    use std::str::FromStr as _;

    use super::*;

    use temp_testdir::TempDir;

    fn args(query: ArgsQuery) -> (common::cli::Args, Args, TempDir) {
        let temp = TempDir::default();
        let common = common::cli::Args {
            verbose: clap_verbosity_flag::Verbosity::new(1, 0),
        };
        let args = Args {
            path_rocksdb: String::from("tests/clinvar-sv/clinvar-sv-grch37.tsv.db"),
            cf_name: String::from("clinvar_sv"),
            cf_name_by_rcv: String::from("clinvar_sv_by_rcv"),
            out_file: temp.join("out").to_string_lossy().to_string(),
            out_format: common::cli::OutputFormat::Jsonl,
            query,
        };

        (common, args, temp)
    }

    #[test]
    fn smoke_query_var_vcv() -> Result<(), anyhow::Error> {
        let (common, args, _temp) = args(ArgsQuery {
            accession: Some("VCV000057688".into()),
            ..Default::default()
        });
        run(&common, &args)?;
        let out_data = std::fs::read_to_string(&args.out_file)?;
        insta::assert_snapshot!(&out_data);

        Ok(())
    }

    #[test]
    fn smoke_query_var_rcv() -> Result<(), anyhow::Error> {
        let (common, args, _temp) = args(ArgsQuery {
            accession: Some("RCV000051426".into()),
            ..Default::default()
        });
        run(&common, &args)?;
        let out_data = std::fs::read_to_string(&args.out_file)?;
        insta::assert_snapshot!(&out_data);

        Ok(())
    }

    #[test]
    fn smoke_query_var_all() -> Result<(), anyhow::Error> {
        let (common, args, _temp) = args(ArgsQuery {
            all: true,
            ..Default::default()
        });
        run(&common, &args)?;
        let out_data = std::fs::read_to_string(&args.out_file)?;
        insta::assert_snapshot!(&out_data);

        Ok(())
    }

    #[test]
    fn smoke_query_var_range_exact() -> Result<(), anyhow::Error> {
        let (common, args, _temp) = args(ArgsQuery {
            range: Some(spdi::Range::from_str("GRCh37:22:34150132:34182300")?),
            ..Default::default()
        });
        run(&common, &args)?;
        let out_data = std::fs::read_to_string(&args.out_file)?;
        insta::assert_snapshot!(&out_data);

        Ok(())
    }

    #[test]
    fn smoke_query_var_range_overlap() -> Result<(), anyhow::Error> {
        let (common, args, _temp) = args(ArgsQuery {
            range: Some(spdi::Range::from_str("GRCh37:22:34150132:34150200")?),
            ..Default::default()
        });
        run(&common, &args)?;
        let out_data = std::fs::read_to_string(&args.out_file)?;
        insta::assert_snapshot!(&out_data);

        Ok(())
    }

    #[test]
    fn smoke_query_var_range_no_overlap() -> Result<(), anyhow::Error> {
        let (common, args, _temp) = args(ArgsQuery {
            range: Some(spdi::Range::from_str("GRCh37:22:34182000:34182300")?),
            ..Default::default()
        });
        run(&common, &args)?;
        let out_data = std::fs::read_to_string(&args.out_file)?;
        insta::assert_snapshot!(&out_data);

        Ok(())
    }
}
