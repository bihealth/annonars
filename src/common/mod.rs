//! Common and utility code.

use std::path::{Path, PathBuf};

pub mod cli;
pub mod keys;
pub mod noodles;
pub mod spdi;

/// The version of `annonars` package.
#[cfg(not(test))]
const VERSION: &str = env!("CARGO_PKG_VERSION");

/// This allows us to override the version to `0.0.0` in tests.
pub fn version() -> &'static str {
    #[cfg(test)]
    return "0.0.0";
    #[cfg(not(test))]
    return VERSION;
}

fn _readlink_f(path: &Path, level: u32) -> Result<PathBuf, anyhow::Error> {
    if level == 0 {
        anyhow::bail!(
            "maximal recursion depth for {:?}",
            path.as_os_str().to_string_lossy()
        )
    } else if path.is_symlink() {
        _readlink_f(&std::fs::read_link(path)?, level.saturating_sub(1))
    } else {
        Ok(path.to_path_buf())
    }
}

/// Recursively resolve the `path`.
pub fn readlink_f<P>(path: P) -> Result<PathBuf, anyhow::Error>
where
    P: AsRef<Path>,
{
    _readlink_f(path.as_ref(), 20)
}
