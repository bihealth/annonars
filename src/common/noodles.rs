//! Noodles utility code.

use std::str::FromStr;

use noodles_vcf::record::info::field;

/// Extract a `String` field from a record.
pub fn get_string(record: &noodles_vcf::Record, name: &str) -> Result<String, anyhow::Error> {
    if let Some(Some(field::Value::String(v))) = record.info().get(&field::Key::from_str(name)?) {
        Ok(v.to_string())
    } else if let Some(Some(field::Value::Array(field::value::Array::String(vs)))) =
        record.info().get(&field::Key::from_str(name)?)
    {
        Ok(vs.first().unwrap().as_ref().unwrap().to_string())
    } else {
        anyhow::bail!("missing INFO/{} in gnomAD record", name)
    }
}

/// Extract a flag field from a record.
pub fn get_flag(record: &noodles_vcf::Record, name: &str) -> Result<bool, anyhow::Error> {
    Ok(matches!(
        record.info().get(&field::Key::from_str(name)?),
        Some(Some(field::Value::Flag))
    ))
}

/// Extract an `i32` field from a record.
pub fn get_i32(record: &noodles_vcf::Record, name: &str) -> Result<i32, anyhow::Error> {
    if let Some(Some(field::Value::Integer(v))) = record.info().get(&field::Key::from_str(name)?) {
        Ok(*v)
    } else if let Some(Some(field::Value::Array(field::value::Array::Integer(vs)))) =
        record.info().get(&field::Key::from_str(name)?)
    {
        Ok(vs.first().unwrap().unwrap())
    } else {
        anyhow::bail!("missing INFO/{} in gnomAD record", name)
    }
}

/// Extract an `f32` field from a record.
pub fn get_f32(record: &noodles_vcf::Record, name: &str) -> Result<f32, anyhow::Error> {
    if let Some(Some(field::Value::Float(v))) = record.info().get(&field::Key::from_str(name)?) {
        Ok(*v)
    } else if let Some(Some(field::Value::Array(field::value::Array::Float(vs)))) =
        record.info().get(&field::Key::from_str(name)?)
    {
        Ok(vs.first().unwrap().unwrap())
    } else {
        anyhow::bail!("missing INFO/{} in gnomAD record", name)
    }
}

/// Extract an `Vec<String>` field from record with an array field.
///
/// This is different than parsing the histograms from pipe-separated strings.
pub fn get_vec_str(record: &noodles_vcf::Record, name: &str) -> Result<Vec<String>, anyhow::Error> {
    if let Some(Some(field::Value::Array(field::value::Array::String(vs)))) =
        record.info().get(&field::Key::from_str(name)?)
    {
        Ok(vs.iter().flatten().cloned().collect())
    } else {
        anyhow::bail!("missing INFO/{} in gnomAD record", name)
    }
}

/// Extract an `Vec<i32>` field from record with an array field.
///
/// This is different than parsing the histograms from pipe-separated strings.
pub fn get_vec_i32(record: &noodles_vcf::Record, name: &str) -> Result<Vec<i32>, anyhow::Error> {
    if let Some(Some(field::Value::Array(field::value::Array::Integer(vs)))) =
        record.info().get(&field::Key::from_str(name)?)
    {
        Ok(vs.iter().flatten().cloned().collect())
    } else {
        anyhow::bail!("missing INFO/{} in gnomAD record", name)
    }
}

/// Extract an `Vec<FromStr>` field from a record encoded as a pipe symbol separated string.
pub fn get_vec<T>(record: &noodles_vcf::Record, name: &str) -> Result<Vec<T>, anyhow::Error>
where
    T: FromStr,
{
    if let Some(Some(field::Value::String(v))) = record.info().get(&field::Key::from_str(name)?) {
        v.split('|')
            .map(|s| s.parse())
            .collect::<Result<Vec<_>, _>>()
            .map_err(|_| anyhow::anyhow!("failed to parse INFO/{} as Vec<_>", name))
    } else {
        anyhow::bail!("missing INFO/{} in gnomAD record", name)
    }
}

/// Extract an `Vec<Vec<FromStr>>` field from a record encoded as a list of pipe symbol
/// separated string.
pub fn get_vec_vec<T>(record: &noodles_vcf::Record, name: &str) -> Result<Vec<T>, anyhow::Error>
where
    T: FromStr,
{
    if let Some(Some(field::Value::Array(field::value::Array::String(value)))) =
        record.info().get(&field::Key::from_str(name)?)
    {
        Ok(value
            .iter()
            .map(|s| {
                s.as_ref()
                    .ok_or(anyhow::anyhow!("missing value in INFO/hap_hl_hist"))
                    .map(|s| {
                        s.split('|')
                            .map(|s| s.parse())
                            .collect::<Result<Vec<T>, _>>()
                    })
            })
            .map(|r| match r {
                Ok(Ok(v)) => Ok(v),
                _ => anyhow::bail!("missing value in INFO/hap_hl_hist"),
            })
            .collect::<Result<Vec<_>, _>>()
            .map_err(|e| anyhow::anyhow!("failed to parse INFO/{} as Vec<Vec<_>>: {}", name, e))?
            .into_iter()
            .flatten()
            .collect())
    } else {
        anyhow::bail!("missing INFO/{} in gnomAD record", name)
    }
}
