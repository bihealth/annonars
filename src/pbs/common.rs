//! Code generate for protobufs by `prost-build`.

/// Information about versions.
pub mod versions {
    include!(concat!(env!("OUT_DIR"), "/annonars.common.versions.rs"));
    include!(concat!(
        env!("OUT_DIR"),
        "/annonars.common.versions.serde.rs"
    ));
}
