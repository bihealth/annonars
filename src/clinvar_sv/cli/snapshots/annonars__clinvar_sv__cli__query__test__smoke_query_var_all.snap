---
source: src/clinvar_sv/cli/query.rs
expression: "&out_data"
---
{"release":"GRCh37","chromosome":"X","start":155210040,"stop":155242832,"innerStart":155210040,"outerStart":155242832,"vcv":"VCV000057500","referenceAssertions":[{"rcv":"RCV000051210","title":"GRCh38/hg38 Xq28(chrX:155980375-156013167)x0 AND See cases","clinicalSignificance":"CLINICAL_SIGNIFICANCE_LIKELY_BENIGN","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"7","start":64691936,"stop":64866073,"innerStart":64691936,"outerStart":64866073,"vcv":"VCV000057566","referenceAssertions":[{"rcv":"RCV000051294","title":"GRCh38/hg38 7q11.21(chr7:65231558-65401160)x1 AND See cases","clinicalSignificance":"CLINICAL_SIGNIFICANCE_LIKELY_PATHOGENIC","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"22","start":34150132,"stop":34182300,"innerStart":34150132,"outerStart":34182300,"vcv":"VCV000057627","referenceAssertions":[{"rcv":"RCV000051361","title":"GRCh38/hg38 22q12.3(chr22:33754145-33786313)x1 AND See cases","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"X","start":73055208,"stop":73059460,"innerStart":73055208,"outerStart":73059460,"vcv":"VCV000057653","referenceAssertions":[{"rcv":"RCV000051388","title":"GRCh38/hg38 Xq13.2(chrX:73835373-73839625)x1 AND See cases","clinicalSignificance":"CLINICAL_SIGNIFICANCE_LIKELY_PATHOGENIC","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"X","start":78605903,"stop":79090819,"innerStart":78605903,"outerStart":79090819,"vcv":"VCV000057654","referenceAssertions":[{"rcv":"RCV000051389","title":"GRCh38/hg38 Xq21.1(chrX:79350406-79835322)x0 AND See cases","clinicalSignificance":"CLINICAL_SIGNIFICANCE_LIKELY_PATHOGENIC","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"X","start":89330088,"stop":89853974,"innerStart":89330088,"outerStart":89853974,"vcv":"VCV000057686","referenceAssertions":[{"rcv":"RCV000051424","title":"GRCh38/hg38 Xq21.31(chrX:90075089-90598975)x1 AND See cases","clinicalSignificance":"CLINICAL_SIGNIFICANCE_LIKELY_PATHOGENIC","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"X","start":102885179,"stop":102967225,"innerStart":102885179,"outerStart":102967225,"vcv":"VCV000057688","referenceAssertions":[{"rcv":"RCV000051426","title":"GRCh38/hg38 Xq22.2(chrX:103630251-103712297)x1 AND See cases","clinicalSignificance":"CLINICAL_SIGNIFICANCE_LIKELY_PATHOGENIC","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"1","start":196757278,"stop":196796716,"innerStart":196757278,"outerStart":196796716,"vcv":"VCV000058538","referenceAssertions":[{"rcv":"RCV000052307","title":"GRCh38/hg38 1q31.3(chr1:196788148-196827586)x1 AND See cases","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"7","start":27194426,"stop":27461954,"innerStart":27194426,"outerStart":27461954,"vcv":"VCV000058544","referenceAssertions":[{"rcv":"RCV000052313","title":"GRCh38/hg38 7p15.2(chr7:27154807-27422335)x1 AND See cases","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"X","start":8595045,"stop":8704378,"innerStart":8595045,"outerStart":8704378,"variantType":"VARIANT_TYPE_DELETION","vcv":"VCV000058565","referenceAssertions":[{"rcv":"RCV000052334","title":"GRCh38/hg38 Xp22.31(chrX:8627004-8736337)x2 AND See cases","clinicalSignificance":"CLINICAL_SIGNIFICANCE_LIKELY_PATHOGENIC","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"X","start":8596894,"stop":8705494,"innerStart":8596894,"outerStart":8705494,"variantType":"VARIANT_TYPE_DELETION","vcv":"VCV000058566","referenceAssertions":[{"rcv":"RCV000052337","title":"GRCh38/hg38 Xp22.31(chrX:8628853-8737453)x2 AND See cases","clinicalSignificance":"CLINICAL_SIGNIFICANCE_LIKELY_PATHOGENIC","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"X","start":8597747,"stop":8674356,"innerStart":8597747,"outerStart":8674356,"variantType":"VARIANT_TYPE_DELETION","vcv":"VCV000058567","referenceAssertions":[{"rcv":"RCV000052338","title":"GRCh38/hg38 Xp22.31(chrX:8629706-8706315)x3 AND See cases","clinicalSignificance":"CLINICAL_SIGNIFICANCE_LIKELY_PATHOGENIC","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_CONFLICTING_INTERPRETATIONS"}]}
{"release":"GRCh37","chromosome":"X","start":28777940,"stop":29112528,"innerStart":28777940,"outerStart":29112528,"variantType":"VARIANT_TYPE_DELETION","vcv":"VCV000058611","referenceAssertions":[{"rcv":"RCV000052386","title":"GRCh38/hg38 Xp21.3(chrX:28759823-29094411)x2 AND See cases","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"X","start":29731388,"stop":29759068,"innerStart":29731388,"outerStart":29759068,"variantType":"VARIANT_TYPE_DELETION","vcv":"VCV000058612","referenceAssertions":[{"rcv":"RCV000052387","title":"GRCh38/hg38 Xp21.2(chrX:29713271-29740951)x3 AND See cases","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"X","start":32680483,"stop":32777081,"innerStart":32680483,"outerStart":32777081,"variantType":"VARIANT_TYPE_DELETION","vcv":"VCV000058614","referenceAssertions":[{"rcv":"RCV000052389","title":"GRCh38/hg38 Xp21.1(chrX:32662366-32758964)x2 AND See cases","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"X","start":105135228,"stop":105669012,"innerStart":105135228,"outerStart":105669012,"variantType":"VARIANT_TYPE_DELETION","vcv":"VCV000058665","referenceAssertions":[{"rcv":"RCV000052442","title":"GRCh38/hg38 Xq22.3(chrX:105891235-106425782)x2 AND See cases","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"Y","start":2651665,"stop":4972987,"innerStart":2651665,"outerStart":4972987,"variantType":"VARIANT_TYPE_DELETION","vcv":"VCV000058795","referenceAssertions":[{"rcv":"RCV000052584","title":"GRCh38/hg38 Yp11.2(chrY:2783624-5104946)x2 AND See cases","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"Y","start":2654967,"stop":2862993,"innerStart":2654967,"outerStart":2862993,"variantType":"VARIANT_TYPE_DELETION","vcv":"VCV000058800","referenceAssertions":[{"rcv":"RCV000052589","title":"GRCh38/hg38 Yp11.2(chrY:2786926-2994952)x2 AND See cases","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"19","start":54104552,"stop":54276282,"innerStart":54104552,"outerStart":54276282,"variantType":"VARIANT_TYPE_DELETION","vcv":"VCV000058805","referenceAssertions":[{"rcv":"RCV000052594","title":"GRCh38/hg38 19q13.42(chr19:53601298-53773028)x3 AND See cases","clinicalSignificance":"CLINICAL_SIGNIFICANCE_LIKELY_PATHOGENIC","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}
{"release":"GRCh37","chromosome":"X","start":8595045,"stop":8704378,"innerStart":8595045,"outerStart":8704378,"variantType":"VARIANT_TYPE_DELETION","vcv":"VCV000144128","referenceAssertions":[{"rcv":"RCV000052335","title":"GRCh38/hg38 Xp22.31(chrX:8627004-8736337)x3 AND See cases","clinicalSignificance":"CLINICAL_SIGNIFICANCE_LIKELY_PATHOGENIC","reviewStatus":"REVIEW_STATUS_CRITERIA_PROVIDED_MULTIPLE_SUBMITTERS_NO_CONFLICTS"}]}

